src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs

```diff
@@ -132,99 +126,128 @@ public virtual byte[] GetValue()
+
+                    asnReadSuccessful = AsnDecoder.TryReadInt32(asnSpan, AsnEncodingRules.BER, out int count, out bytesConsumed);
+                    Debug.Assert(asnReadSuccessful);
+                    asnSpan = asnSpan.Slice(bytesConsumed);
 
-                    int moreData = (int)o[0];
-                    int count = (int)o[1];
-                    byte[] dirsyncCookie = (byte[])o[2];
+                    dirsyncCookie = AsnDecoder.ReadOctetString(asnSpan, AsnEncodingRules.BER, out _);
 
-                    DirSyncResponseControl dirsync = new DirSyncResponseControl(dirsyncCookie, (moreData == 0 ? false : true), count, controls[i].IsCritical, controls[i].GetValue());
+                    DirSyncResponseControl dirsync = new DirSyncResponseControl(dirsyncCookie, moreData != 0, count, controls[i].IsCritical, value);
                     controls[i] = dirsync;
                 }
                 else if (controls[i].Type == "1.2.840.113556.1.4.474")
                 {
-                    // The control is a SortControl.
-                    int result;
+                    // The control is a SortControl, as described in RFC 2891.
+                    ResultCode result;
                     string attribute = null;
-                    object[] o = BerConverter.TryDecode("{ea}", value, out bool decodeSucceeded);
 
-                    // decode might fail as AD for example never returns attribute name, we don't want to unnecessarily throw and catch exception
-                    if (decodeSucceeded)
-                    {
-                        Debug.Assert(o != null && o.Length == 2);
-                        result = (int)o[0];
-                        attribute = (string)o[1];
-                    }
-                    else
+                    AsnDecoder.ReadSequence(asnSpan, AsnEncodingRules.BER, out int sequenceContentOffset, out int sequenceContentLength, out _);
+                    Debug.Assert(sequenceContentLength > 0);
+                    asnSpan = asnSpan.Slice(sequenceContentOffset);
+
+                    result = AsnDecoder.ReadEnumeratedValue<ResultCode>(asnSpan, AsnEncodingRules.BER, out int bytesConsumed);
+                    asnSpan = asnSpan.Slice(bytesConsumed);
+
+                    // Attribute name is optional: AD for example never returns attribute name
+                    asnReadSuccessful = AsnDecoder.TryReadEncodedValue(asnSpan, AsnEncodingRules.BER, out Asn1Tag octetStringTag, out _, out int octetStringLength, out _);
+                    if (asnReadSuccessful)
                     {
-                        // decoding might fail as attribute is optional
-                        o = BerConverter.Decode("{e}", value);
-                        Debug.Assert(o != null && o.Length == 1);
+                        Span<byte> attributeNameBuffer = octetStringLength < scratchSpace.Length ? scratchSpace.Slice(0, octetStringLength) : new byte[octetStringLength];
```

Comment:
nit: use all the bytes available in the scratch buffer
```suggestion
                        Span<byte> attributeNameBuffer = octetStringLength <= scratchSpace.Length ? scratchSpace.Slice(0, octetStringLength) : new byte[octetStringLength];
```